## Part 1: Explore the Data

Import the data and use Pandas to learn more about the dataset.

In [815]:
import pandas as pd

df = pd.read_csv('Resources/client_dataset.csv')
df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number  
0     1096.80     762.71         7.50  105            1  
1       24.95      15.09         1.49   21            0  
2       13.52       7.86         1.68   39            6  
3       36.42      24.85         1.23   29            3  
4      195.10     108.17        46.43   20            1

In [816]:
# View the column names in the data

#print(df.columns)
print("\n".join(df.columns), '\n')

first
last
job
phone
email
client_id
order_id
order_date
order_week
order_year
item_id
category
subcategory
unit_price
unit_cost
unit_weight
qty
line_number 



In [817]:
# Use the describe function to gather some basic statistics
df.describe()

client_id      order_id    order_week    order_year    unit_price  \
count  54639.000000  5.463900e+04  54639.000000  54639.000000  54639.000000   
mean   54837.869416  5.470190e+06     11.359139   2022.993064    136.267207   
std    25487.438231  2.599807e+06      7.023499      0.082997    183.873135   
min    10033.000000  1.000886e+06      1.000000   2022.000000      0.010000   
25%    33593.000000  3.196372e+06      6.000000   2023.000000     20.800000   
50%    53305.000000  5.496966e+06     11.000000   2023.000000     68.310000   
75%    78498.000000  7.733869e+06     17.000000   2023.000000    173.160000   
max    99984.000000  9.998480e+06     52.000000   2023.000000   1396.230000   

          unit_cost   unit_weight           qty   line_number  
count  54639.000000  54639.000000  5.463900e+04  54639.000000  
mean      99.446073      5.004116  5.702646e+02      2.979667  
std      133.164267      5.326599  1.879552e+04      2.436320  
min        0.010000      0.000000  0.000000e+00      0.000000  
25%       14.840000      1.450000  3.200000e+01      1.000000  
50%       49.890000      3.240000  6.800000e+01      3.000000  
75%      125.570000      6.890000  1.700000e+02      5.000000  
max      846.270000     46.430000  3.958244e+06      9.000000

In [818]:
# Use this space to do any additional research
# and familiarize yourself with the data.
print(df.dtypes)


first           object
last            object
job             object
phone           object
email           object
client_id        int64
order_id         int64
order_date      object
order_week       int64
order_year       int64
item_id         object
category        object
subcategory     object
unit_price     float64
unit_cost      float64
unit_weight    float64
qty              int64
line_number      int64
dtype: object


In [819]:
# What three item categories had the most entries?
most_entries_categories = df['category'].value_counts().sort_values(ascending=False)
most_entries_categories.head(3)


category
consumables    23538
furniture      11915
software        8400
Name: count, dtype: int64

In [820]:
# CODES FOR CATEGORY COLUMN WITH ONLY 'CONSUMABLES'
# consumables_df = df.loc[df['category'].str.contains("consumables", na=False)]
# consumables_df = df.loc[df['category'] == 'consumables']
# consumables_df 


In [821]:
# For the category with the most entries, which subcategory had the most entries?
consumables_df = df.loc[df['category'] == 'consumables']
most_entries_subcategories = consumables_df['subcategory'].value_counts().sort_values(ascending=False)
print(f"The subcategory with the most entries in 'consumables' is, {most_entries_subcategories.idxmax()}")
most_entries_subcategories.head()

The subcategory with the most entries in 'consumables' is, bathroom supplies


subcategory
bathroom supplies    6424
paper products       5874
kitchen supplies     5746
misc                 2891
pens                 2603
Name: count, dtype: int64

In [822]:
# Which five clients had the most entries in the data?
df['full_name'] = df['first'] + ' ' + df['last']
frequent_full_names = df['full_name'].value_counts()
frequent_full_names.head(5)

full_name
Jessica Reyes      220
Angela Everett     211
Bryan Myers        209
Alexandra Young    207
Kendra Garrett     207
Name: count, dtype: int64

In [823]:
# Which five clients had the most entries in the data?
most_entries_clients = df['client_id'].value_counts().head(5).sort_values(ascending=False)
most_entries_clients

client_id
33615    220
66037    211
46820    209
38378    207
24741    207
Name: count, dtype: int64

In [824]:
# Store the client ids of those top 5 clients in a list.
most_entries_clients = df['client_id'].value_counts().head(5).index.tolist()
most_entries_clients

[33615, 66037, 46820, 38378, 24741]

In [825]:
df

first       last                      job               phone  \
0         Donald    Harding             Immunologist  793-904-7725x39308   
1        Tiffany      Myers          Music therapist    201.442.4543x942   
2        Shannon     Watson             Immunologist   687.737.9424x8503   
3         Nathan      Baker    Accounting technician    827-788-8123x012   
4      Christina   Schwartz             Chiropractor        265-829-3643   
...          ...        ...                      ...                 ...   
54634     Lauren      Reese           Radio producer  977.877.5272x11382   
54635    Derrick      Moore                 Musician        358.661.5483   
54636     Monica  Gutierrez         Graphic designer    294.805.9100x339   
54637      Wanda    Solomon             Toxicologist       (311)767-4924   
54638     Dwayne       Bell  Chief Financial Officer   606-619-7667x9508   

                                     email  client_id  order_id  order_date  \
0         harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1        myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                    swatson8146@payne.net      57113   1902144  2023-01-29   
3                   bakernathan@benson.com      46554   9031802  2023-04-25   
4      christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   
...                                    ...        ...       ...         ...   
54634   late_reese_4081@montoya-chavez.org      29043   9021716  2023-04-26   
54635         derrick.moore.2602@pope.info      41908   6290153  2023-02-08   
54636            gutierrezm3195@morris.org      35176   8692622  2023-03-05   
54637            solomonwanda5962@ross.org      24485   7592730  2023-01-18   
54638                     belld@galvan.com      24007   7489403  2023-01-04   

       order_week  order_year         item_id     category  \
0              17        2023  EUD29711-63-6U        decor   
1              20        2023  XDA18116-89-4A  consumables   
2               4        2023  ABE59463-05-7E     software   
3              17        2023  ZMM00836-65-0C  consumables   
4              21        2023  BZX55559-12-3X  consumables   
...           ...         ...             ...          ...   
54634          17        2023  UCA28532-31-0B     software   
54635           6        2023  EBD29277-23-7U  consumables   
54636           9        2023  EBD30093-55-9Z  consumables   
54637           3        2023  XMZ38239-49-3D    furniture   
54638           1        2023  ZCZ05589-50-5D     software   

              subcategory  unit_price  unit_cost  unit_weight   qty  \
0                wall art     1096.80     762.71         7.50   105   
1                    pens       24.95      15.09         1.49    21   
2      project management       13.52       7.86         1.68    39   
3                    pens       36.42      24.85         1.23    29   
4                    misc      195.10     108.17        46.43    20   
...                   ...         ...        ...          ...   ...   
54634           logistics       83.13      51.60         2.25    33   
54635    kitchen supplies      206.59     175.46        11.70    47   
54636   bathroom supplies       65.66      57.31         4.16   475   
54637              stands        1.48       1.22        18.04   112   
54638  project management        3.01       1.61         2.07  1031   

       line_number           full_name  
0                1      Donald Harding  
1                0       Tiffany Myers  
2                6      Shannon Watson  
3                3        Nathan Baker  
4                1  Christina Schwartz  
...            ...                 ...  
54634            8        Lauren Reese  
54635            0       Derrick Moore  
54636            6    Monica Gutierrez  
54637            0       Wanda Solomon  
54638            5         Dwayne Bell  

[54639 rows x 19 columns]

In [826]:
# How many total units (the qty column) did the client with the most entries order order?
total_units_ordered_df = df.loc[df['client_id'] == 33615, 'qty'].sum()
total_units_ordered_df


64313

## Part 2: Transform the Data
Do we know that this client spent the more money than client 66037? If not, how would we find out? Transform the data using the steps below to prepare it for analysis.

In [827]:
# Create a column that calculates the subtotal for each line using the unit_price and the qty
import pandas as pd

df['subtotal'] = df['unit_price'] * df['qty']
# df[['subtotal']]


In [828]:
# Create a column for shipping price.
# Assume a shipping price of $7 per pound for orders over 50 pounds and $10 per pound for items 50 pounds or under.
import pandas as pd
import numpy as np

# df['shipping price'] = np.where(df['total_weight'] > 50, df['total_weight'] * 7, df['total_weight'] * 10)
df['total_weight'] = df['qty'] * df['unit_weight']
def shipping_calculator(weight):
    if weight > 50:
        return weight * 7
    return weight * 10

df['shipping_price'] = df['total_weight'].apply(shipping_calculator)
df[['shipping_price', 'total_weight', 'qty', 'unit_weight', 'unit_price', 'subtotal']]



shipping_price  total_weight   qty  unit_weight  unit_price   subtotal
0             5512.50        787.50   105         7.50     1096.80  115164.00
1              312.90         31.29    21         1.49       24.95     523.95
2              458.64         65.52    39         1.68       13.52     527.28
3              356.70         35.67    29         1.23       36.42    1056.18
4             6500.20        928.60    20        46.43      195.10    3902.00
...               ...           ...   ...          ...         ...        ...
54634          519.75         74.25    33         2.25       83.13    2743.29
54635         3849.30        549.90    47        11.70      206.59    9709.73
54636        13832.00       1976.00   475         4.16       65.66   31188.50
54637        14143.36       2020.48   112        18.04        1.48     165.76
54638        14939.19       2134.17  1031         2.07        3.01    3103.31

[54639 rows x 6 columns]

In [829]:
print(df.columns)

Index(['first', 'last', 'job', 'phone', 'email', 'client_id', 'order_id',
       'order_date', 'order_week', 'order_year', 'item_id', 'category',
       'subcategory', 'unit_price', 'unit_cost', 'unit_weight', 'qty',
       'line_number', 'full_name', 'subtotal', 'total_weight',
       'shipping_price'],
      dtype='object')


In [830]:
# Create a column for the total price using the subtotal and the shipping price along with a sales tax of 9.25%
import pandas as pd

df['total_price'] = (df['subtotal'] + df['shipping_price']) * (1 + (9.25 / 100))
df[['total_price', 'subtotal', 'shipping_price']].head()


total_price   subtotal  shipping_price
0  131839.076250  115164.00         5512.50
1     914.258625     523.95          312.90
2    1077.117600     527.28          458.64
3    1543.571400    1056.18          356.70
4   11364.403500    3902.00         6500.20

In [831]:
# Create a column for the cost of each line using unit cost, qty, and
# shipping price (assume the shipping cost is exactly what is charged to the client).
import pandas as pd

df['cost_of_each_line'] = (df['unit_cost'] * df['qty'] + df['shipping_price'])
print(df.columns) 

Index(['first', 'last', 'job', 'phone', 'email', 'client_id', 'order_id',
       'order_date', 'order_week', 'order_year', 'item_id', 'category',
       'subcategory', 'unit_price', 'unit_cost', 'unit_weight', 'qty',
       'line_number', 'full_name', 'subtotal', 'total_weight',
       'shipping_price', 'total_price', 'cost_of_each_line'],
      dtype='object')


In [832]:
# Create a column for the profit of each line using line cost and line price
import pandas as pd

df['profit_of_each_line'] = df['total_price'] - df['cost_of_each_line']
print(df.columns)


Index(['first', 'last', 'job', 'phone', 'email', 'client_id', 'order_id',
       'order_date', 'order_week', 'order_year', 'item_id', 'category',
       'subcategory', 'unit_price', 'unit_cost', 'unit_weight', 'qty',
       'line_number', 'full_name', 'subtotal', 'total_weight',
       'shipping_price', 'total_price', 'cost_of_each_line',
       'profit_of_each_line'],
      dtype='object')


## Part 3: Confirm your work
You have email receipts showing that the total prices for 3 orders. Confirm that your calculations match the receipts. Remember, each order has multiple lines.

Order ID 2742071 had a total price of \$152,811.89

Order ID 2173913 had a total price of \$162,388.71

Order ID 6128929 had a total price of \$923,441.25


In [833]:
# Check your work using the totals above

order_ids = [2742071, 2173913, 6128929]
totals = df[df['order_id'].isin(order_ids)].groupby('order_id')['total_price'].sum()
formatted_totals = totals.apply(lambda x: f'{x:,.2f}')
formatted_totals



order_id
2173913    162,388.72
2742071    152,811.90
6128929    923,441.24
Name: total_price, dtype: object

## Part 4: Summarize and Analyze
Use the new columns with confirmed values to find the following information.

In [834]:
selected_columns = df.loc[:, ['order_id', 'total_price']]
print(selected_columns)

       order_id    total_price
0       8953482  131839.076250
1       8069089     914.258625
2       1902144    1077.117600
3       9031802    1543.571400
4       1322274   11364.403500
...         ...            ...
54634   9021716    3564.871200
54635   6290153   14813.240275
54636   8692622   49184.896250
54637   7592730   15632.713600
54638   7489403   19711.431250

[54639 rows x 2 columns]


In [835]:
# How much did each of the top 5 clients by quantity spend? Check your work from Part 1 for client ids.

filtered_df = df[df['client_id'].isin(most_entries_clients)]
top_5_clients_expenditure = filtered_df.groupby('client_id')['total_price'].sum().reset_index()
top_5_clients_expenditure = top_5_clients_expenditure.merge(df[['first', 'last', 'client_id']],on='client_id', how='left').drop_duplicates()
top_5_clients_expenditure = top_5_clients_expenditure.sort_values(by='total_price', ascending=False)
top_5_clients_expenditure = top_5_clients_expenditure[['first', 'last', 'client_id', 'total_price']]
top_5_clients_expenditure


first     last  client_id   total_price
0       Kendra  Garrett      24741  8.226889e+07
427  Alexandra    Young      38378  1.290655e+07
843     Angela  Everett      66037  1.025951e+07
634      Bryan    Myers      46820  9.743794e+06
207    Jessica    Reyes      33615  8.377309e+06

In [836]:
print(df.columns)

Index(['first', 'last', 'job', 'phone', 'email', 'client_id', 'order_id',
       'order_date', 'order_week', 'order_year', 'item_id', 'category',
       'subcategory', 'unit_price', 'unit_cost', 'unit_weight', 'qty',
       'line_number', 'full_name', 'subtotal', 'total_weight',
       'shipping_price', 'total_price', 'cost_of_each_line',
       'profit_of_each_line'],
      dtype='object')


In [837]:
# Create a summary DataFrame showing the totals for the top 5 clients with the following information:
# total units purchased, total shipping price, total revenue, and total profit. 
# total units purchased = qty * item_id

import pandas as pd
summary_dataframe = df[df['client_id'].isin(most_entries_clients)]
summary_dataframe = summary_dataframe.groupby('client_id').agg({'qty': 'sum',
    'shipping_price': 'sum',
    'cost_of_each_line': 'sum',
    'profit_of_each_line': 'sum'
}).reset_index()
summary_dataframe


client_id     qty  shipping_price  cost_of_each_line  profit_of_each_line
0      24741  239862      5126448.37        45688899.71         3.657999e+07
1      33615   64313      1828984.89         6175313.91         2.201995e+06
2      38378   73667      3429455.40         9634720.98         3.271830e+06
3      46820   75768      1601448.84         7007191.64         2.736603e+06
4      66037   43018      1395151.85         7004482.98         3.255032e+06

In [838]:
summary_dataframe.dtypes

client_id                int64
qty                      int64
shipping_price         float64
cost_of_each_line      float64
profit_of_each_line    float64
dtype: object

In [839]:
# Format the data and rename the columns to names suitable for presentation.
import pandas as pd

summary_dataframe.rename(columns={'client_id': 'client_ID', 'qty': 'quantity', 'shipping_price': '$total_shipping_price',
                                 'cost_of_each_line': '$total_revenue', 'profit_of_each_line': '$total_profit'},inplace=True)



# Define the money columns. 
summary_dataframe.dtypes
# Define a function that converts a dollar amount to millions.
def conversion_to_millions(dollars):
        if isinstance(dollars, (float)):
            return dollars / 1000000
        else:
            return None
        
    
        
# Apply the currency_format_millions function to only the money columns. 
money_columns = ['$total_shipping_price', '$total_revenue', '$total_profit']
summary_dataframe[money_columns] = summary_dataframe[money_columns].applymap(conversion_to_millions)
summary_dataframe

# Rename the columns to reflect the change in the money format. 
summary_dataframe.rename(columns={'$total_shipping_price': '$total_shipping_price(millions)',
                                 '$total_revenue': '$total_revenue(millions)', '$total_profit': '$total_profit(millions)'},inplace=True)
summary_dataframe



C:\Users\denno\AppData\Local\Temp\ipykernel_13552\2627413426.py:22: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  summary_dataframe[money_columns] = summary_dataframe[money_columns].applymap(conversion_to_millions)


client_ID  quantity  $total_shipping_price(millions)  \
0      24741    239862                         5.126448   
1      33615     64313                         1.828985   
2      38378     73667                         3.429455   
3      46820     75768                         1.601449   
4      66037     43018                         1.395152   

   $total_revenue(millions)  $total_profit(millions)  
0                 45.688900                36.579992  
1                  6.175314                 2.201995  
2                  9.634721                 3.271830  
3                  7.007192                 2.736603  
4                  7.004483                 3.255032

In [843]:
# Sort the updated data by "Total Profit (millions)" form highest to lowest and assign the sort to a new DatFrame.
sorted_by_millions = summary_dataframe.sort_values(by='$total_profit(millions)', ascending=False)
sorted_by_millions

client_ID  quantity  $total_shipping_price(millions)  \
0      24741    239862                         5.126448   
2      38378     73667                         3.429455   
4      66037     43018                         1.395152   
3      46820     75768                         1.601449   
1      33615     64313                         1.828985   

   $total_revenue(millions)  $total_profit(millions)  
0                 45.688900                36.579992  
2                  9.634721                 3.271830  
4                  7.004483                 3.255032  
3                  7.007192                 2.736603  
1                  6.175314                 2.201995

Through the analysis of this huge dataset, I have been able to find crucial information that would be very important in learning about the clients, their popular product categories, the top and the bottom customers and so much more.
I have also been able to explore and manipulate the data to get very specific information throughout the dataset.
Finally, I have been able to perform various, very targeted calculations to the dataset while formatting it to a presentable format which would be easy to present.